### ISSUE TITLE
Pure parsing of Dhall files

### ISSUE THREAD

**Issue by srid** [dhall-lang/dhall-haskell#354]
I'm trying to parse my Dhall files from a Reflex frontend (running all in web browser), and for this I need a way to use Dhall outside of its usual IO actions.

Is there a way to use Dhall parsing (with creation of Haskell types) in a pure fashion without involving the IO monad?

One challenge would be how to resolve linked files (eg: `./Foo.md as Text`). And perhaps it would be fitting to solve it lazily (not loaded right away), with a future handler that does the actual loading (basically a Reflex route that fetches the file from the server).

Thoughts? This would [allow me](https://github.com/srid/deux/projects/1) to write a blog engine that runs entirely on the frontend, without using any database, but only .dhall and .md files. I believe this would be a compelling app using Dhall. :-)

**Comment by Gabriel439** [comment]
Yes, this is possible using the current API with some effort if you connect the following steps:

*   Parsing is already pure using `Dhall.Parser.exprFromText`

*   Import resolution is pure if you use `Dhall.Import.loadWith`

    *   Also, if your program has no imports, you can simplify this step using:

        ```haskell
        traverse (\_ -> Nothing) :: Expr s Path -> Maybe (Expr s X)
        ```

        That will return a `Just` result if the expression was import-free

*   Type-checking is pure using `Dhall.TypeCheck.typeOf`

*   Normalization is pure using `Dhall.Core.normalize`

*   Converting the value to Haskell is pure using `Dhall.extract Dhall.auto`

Basically, import resolution is the part that makes `input` impure, but if you swap it out with a pure equivalent (or don't resolve imports at all) then the entire process is pure

**Comment by Gabriel439** [comment]
I will go ahead and close this since I believe the question was answered
